src/engine/profiler.h

```diff
@@ -94,81 +254,816 @@ struct DevStat {
+    general_stats_.opr_exec_stats_->enqueue(stat->release());
+  }
+
   /*! \brief generate device information following chrome profile file format */
-  void EmitPid(std::ostream *os, const std::string& name, uint32_t pid);
-  /*! \brief generate event information following chrome profile file format */
-  void EmitEvent(std::ostream *os, const std::string& name,
-          const std::string& category, const std::string& ph,
-          uint64_t ts, uint32_t pid, uint32_t tid);
+  void EmitPid(std::ostream *os, const std::string& name, size_t pid);
+
   /*! \brief Profiler instance */
   static Profiler* instance_;
   /*! \brief internal mutex of the profiler */
-  std::mutex m_;
+  std::recursive_mutex m_;
   /*! \brief indicate whether the profiler is running */
   ProfilerState state_;
   /*! \brief once running, enable profiler to output */
   bool enable_output_;
   /*! \brief indicate what operator the profiler will record */
-  ProfilerMode mode_;
+  int mode_;
   /*! \brief filename to output profile file */
   std::string filename_;
   /*! \brief profile statistics consist of multiple device statistics */
-  DevStat* profile_stat;
+  DeviceStats* profile_stat;
+  /*! \brief Stats not associated directly with a device */
+  DeviceStats  general_stats_;
+  /*! \brief Map category -> pid */
+  std::unordered_map<std::string, size_t> category_to_pid_;
   /*! \brief cpu number on the machine */
   unsigned int cpu_num_;
   /*! \brief gpu number on the machine */
   unsigned int gpu_num_;
   /*! \brief the profiler init time */
   uint64_t init_time_;
+  /*! \brief Last profile append mode */
+  volatile bool append_profile_ = false;
+  /*! \brief Number of non-meta profiling record emitted */
+  volatile uint64_t num_records_emitted_ = 0;
+  /*! \brief Number of times profile was dumped */
+  volatile uint64_t profile_dump_count_;
+  /*! \brief Whether profiling is paused */
+  volatile bool paused_ = false;
+  /*! \brief Asynchronous operation thread lifecycly control object */
+  std::shared_ptr<dmlc::ThreadGroup> thread_group_ = std::make_shared<dmlc::ThreadGroup>();
+  /* !\brief pids */
+  std::unordered_set<uint32_t> process_ids_;
 };
 
-/*! \return current clock time, time unit is microsecond (10^-6 s) */
-inline uint64_t NowInUsec();
-/*! \brief set operation execution start timestamp */
-void SetOprStart(OprExecStat* opr_stat);
-/*! \brief set operation execution end timestamp */
-void SetOprEnd(OprExecStat* opr_stat);
+#ifdef MXNET_USE_VTUNE
+#define VTUNE_ONLY_CODE(...) __VA_ARGS__  /* This is undefined at the bottom of this file */
+#else
+#define VTUNE_ONLY_CODE(...) /* */        /* This is undefined at the bottom of this file */
+#endif
+
+/**
+ *  _____              __  _  _  _                ____  _     _            _
+ * |  __ \            / _|(_)| |(_)              / __ \| |   (_)          | |
+ * | |__) |_ __  ___ | |_  _ | | _ _ __   __ _  | |  | | |__  _  ___   ___| |_  ___
```

Comment:
did you copy paste this from somewhere? License issues?

Reply:
the comment? It gets generated from my slightly modified version of figlet (I added C/C++ comment support):
https://github.com/cjolivier01/figlet